gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala

```diff
@@ -471,25 +473,37 @@ case class ColumnarOverrideRules(session: SparkSession)
       (_: SparkSession) => FallbackEmptySchemaRelation(),
       (_: SparkSession) => StoreExpandGroupExpression(),
       (_: SparkSession) => AddTransformHintRule(),
-      (_: SparkSession) => TransformPreOverrides(supportAdaptive),
+      (_: SparkSession) => TransformPreOverrides(this.isLeafPlanExchange || this.isAdaptiveContext),
       (_: SparkSession) => RemoveTransformHintRule()) :::
       BackendsApiManager.getSparkPlanExecApiInstance.genExtendedColumnarPreRules()
   }
 
-  def postOverrides(supportAdaptive: Boolean): List[SparkSession => Rule[SparkPlan]] =
-    List((s: SparkSession) => TransformPostOverrides(s, supportAdaptive)) :::
+  def postOverrides(): List[SparkSession => Rule[SparkPlan]] =
+    List((s: SparkSession) => TransformPostOverrides(s, this.isAdaptiveContext)) :::
       BackendsApiManager.getSparkPlanExecApiInstance.genExtendedColumnarPostRules() :::
       List((_: SparkSession) => ColumnarCollapseCodegenStages(GlutenConfig.getConf))
 
   override def preColumnarTransitions: Rule[SparkPlan] = plan => PhysicalPlanSelector.
     maybe(session, plan) {
       var overridden: SparkPlan = plan
       val startTime = System.nanoTime()
-      supportAdaptive = AdaptiveSparkPlanUtil.supportAdaptiveWithExchangeConsidered(plan)
+      this.isLeafPlanExchange = plan match {
+        case _: Exchange => true
+        case _ => false
+      }
+      val traceElements = Thread.currentThread.getStackTrace
```

Comment:
If we have to do this in a hacky way, it's probably better to move the depth counting part into shim layer.

Reply:
Thanks for your comment!
I guess this calling stack is stable. And at least for our currently supported spark versions, there is no difference. I think if in the future other spark versions change it, we can differentiate the impl. in shim layers. Right?

Reply:
Sure. To make the code more readable, maybe we can name the number "13" to something like "AqeStatcktraceDepth"?

Reply:
Yes, just made some changes in the latest commit. Thanks!